### ISSUE TITLE
Incorrect training data

### ISSUE THREAD

**Issue by KlausBuchegger** [tudarmstadt-lt/kaldi-tuda-de#15]
Hi,
so I trained Kaldi using your (old) s5 script, and as a sanity check I tried to decode the training data.
When I compared the text file from the training data to my results, I noticed that there seem to be quite a number of errors in the texts.
I checked the audio and xml files and saw that the sentences were wrong.

I added a screenshot of a partial vimdiff of the text and my results.

![kaldi](https://user-images.githubusercontent.com/14022837/39804304-f54c8f64-5373-11e8-9f36-3509f78478d1.PNG)

Appears to be a mixup, as those sentences do exist, but in other audio files

**Comment by bmilde** [comment]
Hi,
thanks for opening the issue. I'm was aware of this and I'm currently investigating this.

By running
https://github.com/tudarmstadt-lt/kaldi-tuda-de/blob/master/s5_r2/local/run_cleanup_segmentation.sh
from the new set of scripts in s5_r2 a cleaned training directory is generated. I've compared it to the train set and this seem to be the broken ids:

https://github.com/tudarmstadt-lt/kaldi-tuda-de/blob/master/s5_r2/local/cleanup/problematic_wavs.txt

The cleanup script also decodes the training set (with a biased language model built from the reference), so that's similar to what you did. The ids from s5_r2 are similar to s5, there is just a suffix added for the different microphones (a,b,c) since s5 didn't use all the available data.

Can you send me your diff and/or would you be able to fix the corpus XML's directly? Otherwise since this doesn't seem to be a lot of the total data (1.5%), I'd simply remove the broken data from the next release (v3) of the corpus.

**Comment by svenha** [comment]
If fixing the XML files needs some helping hands and makes sense, let me know.

**Comment by bmilde** [comment]
Hi svenha!

That would be greatly appreciated! There is already a list of files with
wrong transcriptions in the cleanup folder in the repository. That should
be a good starting point!

I'm currently preparing v3 of the corpus, where I sorted these files out
into a separate folder. I can send you a link tomorrow.

svenha <notifications@github.com> schrieb am Mo., 28. Mai 2018, 16:17:

**Comment by bmilde** [comment]
Here is our planned v3 package where hopefully most of the bad utterances are moved into a separate folder: http://ltdata1.informatik.uni-hamburg.de/kaldi_tuda_de/german-speechdata-package-v3.tar.gz

@KlausBuchegger can you upload your diff and decoded output?

**Comment by svenha** [comment]
Two questions about the file problematic_wavs.txt.
- It lists 730 files, but the directory train_removed/ contains 363 recordings (5 files per recording). Why?
- What is is the role of the transcript in problematic_wavs.txt? Is it the incorrect transcript from the current .xml file?

I think the uploads from @KlausBuchegger might be helpful ...

**Comment by fbenites** [comment]
Hello 
I am trying to train deepspeech and I got two days ago a feeling it might lies on the data, that the model does not converge well. I checked also the the data and have a rather preliminary impression.
As I see some transcriptions got delete but some got mixed, so the text exists but is (was) assigned to other transcription (or of course the same text was spoken multiple times):
train/2014-03-17-13-03-33_Kinect-Beam.wav should have the same text as train/2014-05-08-11-48-47_Kinect-Beam.wav 

What speaks for mixed (wrong) assignment is that sometimes I have the impression the text was just shifted:

train/2014-03-17-13-03-49_Kinect-Beam.wav should have the same text as train/2014-03-17-13-03-33_Kinect-Beam.wav
and
train/2014-03-17-13-04-15_Kinect-Beam.wav and train/2014-03-17-13-04-43_Kinect-Beam.wav

The dates are pretty close (they are in my training file neighbor lines).
Should we use some sort of work coordination sorting it out?

**Comment by svenha** [comment]
@fbenites Yes, it would be good to distribute the correction work. Benjamin (bmilde) offered to produce a list of problematic transcripts by running the recognizer on the train set. When this is done, let us partition the manual check work in two parts.

**Comment by bmilde** [comment]
@fbenites @svenha thanks for offering to help. I'm running the decode on the train set right now and should have the results soon. 

@svenha As for the number of problematic utterances in the proposed v3 tar: the numbers are different since I excluded all wav files of all microphones, whenever at least one decode of one microphone fails. There are multiple microphone recordings of the same utterance. Better safe than sorry. Still, only about 1.5% of all utterances are problematic. But the problematic files tend to be in the same recording session(s).

@fbenites Low RNN-CTC performance will probably remain, even if we fix all the problematic files. There is probably not enough data for end-to-end RNN training (40h x number of microphones, but that is more like doing augmentation learning on 40h of data). What kind of WERs are you seeing with DeepSpeech at the moment? Are you training with or without a phonetic dictionary? Utterance variance is unfortunately also fairly low, there are only about 3000 distinct sentences in train/dev/test combined, so I suggest using a phoneme dictionary if possible. I also suggesting adding German speech data from SWC (https://nats.gitlab.io/swc/), worked very well in the our r2 scripts in this repository (18.39% WER dev /  19.60% WER test now).

**Comment by fbenites** [comment]
@bmilde Thanks, I will have a look at the wikipedia.
I am not certain, I removed some problematic files I am using also: github.com/ynop/audiomate to process which cover some files already, but I added the other 700 to blacklist.
I will have some results tomorrow. WER is also complicated see https://github.com/ynop/deepspeech-german sometimes the text is just missing some chars or spaces, which drops the WER a lot. I had some useful results only using voxforge contradicting the results with voxforge and tuda. I will check the phoneme dictionary in deepspeech, thx.

**Comment by bmilde** [comment]
@fbenites  @svenha  

I uploaded the decode of the tuda train set to: http://speech.tools/decode_tuda_train.tar.gz

This file might be interesting:

exp/chain_cleaned/tdnn1f_1024_sp_bi/decode_tuda_train/scoring_kaldi/wer_details/per_utt

or alternatively you can also diff e.g.:

diff exp/chain_cleaned/tdnn1f_1024_sp_bi/decode_tuda_train/scoring_kaldi/penalty_0.5/10.txt  exp/chain_cleaned/tdnn1f_1024_sp_bi/decode_tuda_train/scoring_kaldi/test_filt.txt

Though this doesn't look so pretty in standard diff, so some graphical diff-like tool like in the screenshot from Klaus is probably a better idea.

**Comment by svenha** [comment]
Thanks @bmilde . I investigated the file per_utt with a little script that sums the isd part of the "cisd" fields. I count all of them as errors of equal weight. Then I used an error threshold t. For t=4, 2490 affected files; for t=5, 1788 files, for t=6,1511 files. If I ignore the microphone suffix (_a etc.), there remain 616 files to be checked for t=6. As we have only 2 annotators, I would suggest to use t=6. (We can repeat the decoding of the train set with the improved corpus and choose a second annotation round.) If you agree, I can produce the file list, sort it, and cut it in the middle. I would take the first half, @fbenites the second part?

@bmilde : How should we contribute the changes? We could edit the .xml files, collect all changed .xml files and send them to you. But I am open to other approaches, like a normal pull request (if the repo is not too large for this).

One final point: in per_utt, there are file names like 02dae8284f104451a8de85538da6fdec_20140317140355_a . Is it safe to assume that it corresponds 1:1 to an xml file derived from the date/time part, here 2014-03-17-14-03-55.xml ?

**Comment by bmilde** [comment]
Many thanks @svenha !

Yes, it is safe to assume that 02dae8284f104451a8de85538da6fdec_20140317140355_a belongs to 2014-03-17-14-03-55.xml

02dae8284f104451a8de85538da6fdec is the speaker hash, the last letter indicates the microphone, between that the IDs contain the timestamp without the dashes. Note that it's also safe to assume that _a, _b, _c etc all belong to the same utterance and that they should contain the same transcription. If all of them decode to something else, it's safe to assume an incorrect transcription.

Maybe it's also a good idea to make the threshold t depended on the length of the transcription, there are some very short utterances, too. But I can also always rerun the decoding for you after getting corrected xml files.

Since the corpus is not hosted on Github, it's probably easier if you send me the corrected xml files directly. But you can also send me a pull request, containing only the corrected xml files, placed somewhere in a subfolder of the local directory. I can then also write a script that checks the corpus files and patches them if needed, so that it is not necessary to redownload the whole tar.gz file.

One final point: Note that the xml files have sentence IDs. 

E.g. sentence_id 59 in:

```
<?xml version="1.0" encoding="utf-8"?><recording><speaker_id>02dae828-4f10-4451-a8de-85538da6fdec</speaker_id><rate>16000</rate><angle>0</angle><gender>male</gender><ageclass>21-30</ageclass><sentence_id>59</sentence_id><sentence>Rom wurde damit zur ‚De-Facto-Vormacht‘ im ö
stlichen Mittelmeerraum.</sentence><cleaned_sentence>Rom wurde damit zur De Facto Vormacht im östlichen Mittelmeerraum</cleaned_sentence><corpus>WIKI</corpus><muttersprachler>Ja</muttersprachler><bundesland>Hessen</bundesland><sourceurls><url>https://de.wikipedia.org/wiki/R|
ömisches_Reich</url></sourceurls></recording>
```

A text file with all of the sentence IDs and transcriptions is in the root of the corpus archive. Since there are multiple recordings per sentence ID, it is very unlikely that the correct sentence is not included. Maybe we can also just try to find the closest match automatically and check that its correct manually?

**Comment by svenha** [comment]
I switched from an absolute threshold to a relative threshold as suggested by @bmilde  : number_of_errors / number_of_words >= 0.2 
I include an xml file only if all of its recordings (i.e. microphones a, b, c, and d) fufill this criterion.
This gave me 744 xml file names that I attach in two parts of 372 file names each below. I will check the first part (files1) now and send corrected xml files. (We will see how theses changes must be propagated to the SentencesAndIds files.) If the sentence-id was moved by 1 or similar (as noted by others), I will correct the sentence-id and delete the elements sentence and cleaned_sentence.

[tuda-20perc-err.files1.txt](https://github.com/uhh-lt/kaldi-tuda-de/files/2175909/tuda-20perc-err.files1.txt)
[tuda-20perc-err.files2.txt](https://github.com/uhh-lt/kaldi-tuda-de/files/2175910/tuda-20perc-err.files2.txt)

**Comment by svenha** [comment]
tuda-20perc-err.files1.txt is finished; my manual correction speed was around 2 minutes per recording. I will send the files to @bmilde. Any volunteers for tuda-20perc-err.files2.txt? (If not, I might have time in August.)

**Comment by svenha** [comment]
Just to avoid duplicate work, I would let you know that I am working on files2.

**Comment by svenha** [comment]
The second half (i.e. files2) were finished some weeks ago. Is there a rough estimate of the release date of tuda v3?

**Comment by silenterus** [comment]
Hey guys, huge thanks for the dataset!!
i was able to combine your dateset, voxforge, librivox and the swc corpus. 
I put much effort in cleaning the datasets to make sure there are as little mistakes as possible. 

I trained it with the deepspeech framework and 150.000 samples and got 12% WER.
Since then i corrected some other mistakes and fixed some more errors. 
I just started the second round with 220.000 clean samples. 
I hope i will get under 10% WER. 

I will share the code and the results after the training.

Greets from Frankfurt :P

**Comment by silenterus** [comment]
PS : 
cat all_corpora_filtered_maryfied.txt | grep " hhe"

**Comment by wolfgang-s** [comment]
Great work @silenterus 

Do you have an update on the WER?

Do you plan on including data from common voice?  They claim to already have 127 h of validated german voice https://voice.mozilla.org/

**Comment by silenterus** [comment]
Okay sry to keep you guys so long wating.
First the bad news....
The graph had only 16% WER. I was really disappointed and ashamed because i misinterpreted the loss value .
See in all my previously trainings this value correlated with the resulting WER. The loss value was 14% when i reached 12% WER with my previous attempt. 

But the good news i'm ready to share the the code and im pretty sure someone will find the right parameters to archive 10%.
I did implement an Common Voice parser but i couldn't find any German Corpus yet. If someone find one plz share link. 



# Installation :

sudo apt-get install sqlite3 -y


 
### if you want to buld tri and binary (you need to install them otherwise the script wont find them)
git clone https://github.com/kpu/kenlm
cd kenlm
mkdir -p build
cd build
cmake ..
make -j 4
make install      # important !!!
cd ..
cd ..



git clone https://github.com/silenterus/deepspeech-cleaner

cd deepspeech-cleaner

pip3 install -r requirements.txt



### if you want to use all the samples that are currently in my db you have to download these files. (keep traffic humane..it's a private server) 


http://pia.servehttp.com:8100/de_DE.tgz #unchanged
>>> deepspeech-cleaner/datasets/librivox/by_book/

http://pia.servehttp.com:8100/german-speechdata-package-v2.tar.gz #unchanged
>>> deepspeech-cleaner/datasets/tuda/tuda/train
>>> deepspeech-cleaner/datasets/tuda/tuda/dev
>>> deepspeech-cleaner/datasets/tuda/tuda/test


https://pia.servehttp.com:8100/wiki.zip   
>>> deepspeech-cleaner/datasets/wiki/audio/

http://pia.servehttp.com:8100/vox.zip   
>>> deepspeech-cleaner/datasets/wiki/audio/


http://pia.servehttp.com:8100/audio.db
>>> deepspeech-cleaner/

## the server will be online for the next 7 days





### all the parameters for sql query
python3 deepspeech-cleaner.py --help


### creating training files in "training/defaulter" with default values without lm.binary + trie

python3 deepspeech-cleaner.py create_files


### creating training files in "training/german300k" with default values without lm.binary + trie

python3 deepspeech-cleaner.py create_files --training german300k



### creating training files in "training/german200k" with default values and lm.binary + trie
### put all_corpora_filtered_maryfied.txt in 
>>> deepspeech-cleaner/cleaner 
#### the paramater doesnt work yet
#### i used DeepSpeech Version 0.2.1-alpha.2 to generate the trie. I hope they didnt change the language model since then

python3 deepspeech-cleaner.py create_files --training german200k --trie True


### start training in defaulter
bash start_train.sh defaulter


### start training in german300k
bash start_train.sh german300k



## and dont mind the warnings when you preprocessing

#WARNING:root:frame length (551) is greater than FFT size (512), frame will be truncated. Increase NFFT to avoid.
#WARNING:root:frame length (551) is greater than FFT size (512), frame will be truncated. Increase NFFT to avoid.

i did, and even after i red they arent harmful, i insist that i "repair" the wavs. So a resampled every wav. After that i got the 16% WER :P





that's hardly all but enough to start trainning.


i think an offline translator for several languages would be also possible and awesome :D

that's what im trying to archive. to make this git functional for multiple languages.

i working on a voxforge downloader for multiple languages 
the module num2words supports many languages but not without some errors. for example 1. = erster

If you want to help i would be more than thankful. as you might have guessed, my english sucks.

**Comment by fbenites** [comment]
Hi,

Sorry for the radio silence, I was caught up in other projects.
Where are the cleaned data? I would like to check them further. at http://speech.tools/ I get a 403.

Thanks again!

**Comment by silenterus** [comment]
Hey, 
so here are the trie, lm, alphabet and graph from my attempt with upper letters. 

alphabet >>> https://drive.google.com/open?id=1hnINb5NQI6a1kG2COFrluRrkRaa8MRJy
graph >>> https://drive.google.com/open?id=1BQfXIplWQG0FB2PggeDFxzsbKxRNg07P
lm.binary >>> https://drive.google.com/open?id=1eQRowHTZoyHNvp6A4QA1MjotWH3p78XD
trie >>> https://drive.google.com/open?id=1pC0jRXZ6xUHiIxWq43t6EW59vXcuXyaj

i used DeepSpeech Version : 0.4.0-alpha.0


I took 28 h to export the graph. In this time my desktop froze and i couldn't see the resulting WER. 
I tested the graph with a couple of wavs but i didn't evaluate the WER because i wanted to start directly a new training with only lowercase letters. 

But i suppose it's something between 14-17%


Training with lowercase so far : 
I Validation of Epoch 5 - loss: 27.686189                                                               
100% (883 of 883) |###############################################| Elapsed Time: 0:05:53 Time:  0:05:53
I Training epoch 6...
I Training of Epoch 6 - loss: 20.015966                                                                 
100% (8241 of 8241) |#############################################| Elapsed Time: 2:12:24 Time:  2:12:24
I Validating epoch 6...
  2% (20 of 883) |# 


i prepared a git which automatically download and prepare the wavs. I hope nobody wasted the time with the current version. It's not functional.....

I just added the librivox downloader but didnt test it so far.
tuda/voxforge already tested from download till training. 

I maybe i get it to work today and update the git. We will see

**Comment by akoehn** [comment]
@silenterus: It would be nice to keep the discussion on-topic.  This bug is about mix-ups and errors in the data files.  Feel free to open a new bug about training deep speech with our data.

@fbenites: speech.tools is hosted by @bmilde afaik, maybe he can fix that.

**Comment by silenterus** [comment]
Sry you are absolutly right.
I will put the results on my git. 
Keep up the good work

**Comment by svenha** [comment]
Are there any plans to integrate all the corrections from 2018 or later into a new corpus version?

**Comment by Alienmaster** [comment]
I created a repository with the whole ~20GB Dataset here: https://github.com/Alienmaster/TudaDataset
For the revision v4 i removed the incorrect data mentioned [here](https://github.com/uhh-lt/kaldi-tuda-de/blob/master/s5_r2/local/tuda_files_to_skip.txt) and added the corrections made by @svenha .
Currently i train on this dataset without any errors.
Feel free to download and test the new revision.

**Comment by svenha** [comment]
@Alienmaster Thanks for picking this up and the clever issue template in the new TudaDataset repo.

If you have any new evaluation results, please let us know :-)

**Comment by bmilde** [comment]
Closing this, release 4 of the tuda dataset contains the fixes: http://ltdata1.informatik.uni-hamburg.de/kaldi_tuda_de/german-speechdata-package-v4.tar.gz

Our CV7 branch uses this already to train the models, together with the newest Common voice data. See https://github.com/uhh-lt/kaldi-tuda-de/tree/CV7
